util/chunk/row_container.go

```diff
@@ -239,17 +277,200 @@ func (a *SpillDiskAction) SetFallback(fallback memory.ActionOnExceed) {
 // SetLogHook sets the hook, it does nothing just to form the memory.ActionOnExceed interface.
 func (a *SpillDiskAction) SetLogHook(hook func(uint64)) {}
 
-// ResetOnce resets the spill action so that it can be triggered next time.
-func (a *SpillDiskAction) ResetOnce() {
+// ResetRowContainer resets the spill action and sets the RowContainer for the SpillDiskAction.
+func (a *SpillDiskAction) ResetRowContainer(c *RowContainer) {
+	a.m.Lock()
+	defer a.m.Unlock()
+	a.c = c
+}
+
+// WaitForTest waits all goroutine have gone.
+func (a *SpillDiskAction) WaitForTest() {
+	a.testWg.Wait()
+}
+
+// ErrCannotAddBecauseSorted indicate that the SortedRowContainer is sorted and prohibit inserting data.
+var ErrCannotAddBecauseSorted = errors.New("can not add because sorted")
+
+// SortedRowContainer provides a place for many rows, so many that we might want to sort and spill them into disk.
+type SortedRowContainer struct {
+	*RowContainer
+	ptrM struct {
+		// RWMutex guarantees the operator for rowPtrs is mutually exclusive.
+		sync.RWMutex
+		// rowPointer store the chunk index and row index for each row.
+		rowPtrs []RowPtr
+	}
+
+	ByItemsDesc []bool
+	// keyColumns is the column index of the by items.
+	keyColumns []int
+	// keyCmpFuncs is used to compare each ByItem.
+	keyCmpFuncs []CompareFunc
+
+	actionSpill *SortAndSpillDiskAction
+}
+
+// NewSortedRowContainer creates a new SortedRowContainer in memory.
+func NewSortedRowContainer(fieldType []*types.FieldType, chunkSize int, ByItemsDesc []bool,
+	keyColumns []int, keyCmpFuncs []CompareFunc) *SortedRowContainer {
+	return &SortedRowContainer{RowContainer: NewRowContainer(fieldType, chunkSize),
+		ByItemsDesc: ByItemsDesc, keyColumns: keyColumns, keyCmpFuncs: keyCmpFuncs}
+}
+
+// Close close the SortedRowContainer
+func (c *SortedRowContainer) Close() error {
+	c.GetMemTracker().Consume(int64(-8 * cap(c.ptrM.rowPtrs)))
```

Comment:
Should set `c.ptrM.rowPtrs` to `nil` to prevent the unrelease of `SortedRowContainer `.